### DISCUSSION TITLE
Contouring roadmap

### DISCUSSION THREAD

**Discussion by ianthomas23** [bokeh/bokeh#12234] [General]
Now that we have support for contour plots (#12020) here is my roadmap for further work, in roughly priority order.

  - [x] Documentation: user guide page and examples. This is in progress and will be in the 3.0 release. Completed by #12284.
  - [ ] Support requesting a number of contour levels (e.g. `fig.contour(levels=10, ...)`) rather than the levels themselves. The code will choose sensible levels based on the data range of `z`. Initially the levels will be linearly-spaced. The `levels` kwarg will then be optional and the default will be some integer TBD. It won't guarantee you receive that exact number of levels as the requirement that the levels are sensibly spaced is more important.
  - [ ] Logarithmically distributed levels (see above). This information should also be passed on to the colorbar so that it uses an appropriate scale and ticker. This work will define the API for specifying non-linear levels/scale/ticks so that we can support other options in future.
  - [ ] Other ways of specifying vector visual properties when you don't know how many levels you will have, perhaps similar to `linear_palette()`. Example use case would be different line color and dash depending on whether contour level is positive or negative. Also consider use of `color` for both `fill_color` and `line_color` (and related `alpha`) as other glyphs do.
  - [ ] Extending the colormapping/colorbars to color filled polygons outside of the `levels` range in min/low and max/high directions. For example, if you use `levels=[0, 1, 2]` at the moment you'll get 3 contour lines, one at each level, and two sets of filled polygons, the first from `z=0` to `z=1` and the second from `z=1` to `z=2`. When you ask to "extend in max/high direction" (API is TBD) it will calculate and render another set of filled contour polygons for `z=2` up to a `z` of infinity (probably max `z` value) and the top of the colorbar will have a colored triangle to indicate this. See https://matplotlib.org/stable/tutorials/colors/colorbar_only.html for Matplotlib's equivalent.
  - [ ] Automatic placement of text annotations on contour lines to show their `z` levels. This is initially a research project to identify how other libraries do it and come up with a sensible design, and doesn't necessarily have to be done by me :smile:
  - [ ] Possibility of implementing a new data transfer format from ContourPy to Bokeh to avoid too much data wrangling in the Bokeh code.

**Comment by ianthomas23** [comment]
The next item I plan to implement here is to better support for users to specify vector visual properties without knowing the number of levels contoured. This support is required before we relax the `levels` argument to allow levels to be calculated from the supplied data limits.

Currently non-color vector visual properties must be sequences of exactly the right length. `line_color` and `fill_color` are more flexible in that they accept

1. A `PaletteCollection` (dict of `int` to `palette`) such as `Sunset` to lookup the palette based on the number of contours. There are not many palette collections with more than 11 colors, and I don't want to add more just to support contour plots.
2. If passed a palette longer than required then `linear_palette()` is used to resample the palette down to the required size. This works well for the length 256 palettes such as Viridis256 and colorcet palettes.

I propose two enhancements, one for colors and one general purpose for all visual properties.

## Colors only

Rather than use `linear_palette()` to resample fixed-length palettes, I'd like to introduce a new `interp_palette()` function that takes a fixed-length palette and a required length `n` (like `linear_palette`) but it uses floating-point linear interpolation (separately on R, G, B, and A components) to resample to the required length and it can thus resample to both longer and shorter palette lengths, whereas `linear_palette` can only resample to shorter length. 

This functionality could be put in the existing `linear_palette` rather than have a new function but that would change the behaviour of `linear_palette` as it uses the `floor` of the interpolated index rather than color interpolation.

`interp_palette` would be implemented both in Python and in BokehJS as although it is only needed in Python for contouring it is a useful function to have available in BokehJS.

Edit: `interp_palette` added in PR #12789, other palette length options in #13212.

## General

This enhancement is specific to contour plots and does not add any new public functions. If a `Callable` is passed as an argument to a visual property then that `Callable` is called with the sequence of contour levels (actually it will probably be a 1D numpy array of floats) and it will return the vector visual property. An example will help explain:
```python
def line_dash(levels):
    return ["solid" if level >= 0.0 else "dashed" for level in levels]

p.contour(..., line_dash=line_dash)
```
It could be a lambda of course, and some `numpy` functions like `where` could be useful in such callbacks.

This will be on the Python side of Bokeh only.

**Comment by itsgifnotjiff** [comment]
Selfishely I woul love to see the `clabels` implemented into `Bokeh` and as a small reference here is the [Matplotlib `contour` and `clabel` code](https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/contour.py) as well as the respective [documentation](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.clabel.html) and [demo](https://matplotlib.org/stable/gallery/images_contours_and_fields/contour_label_demo.html). Thank you so much Mr. @ianthomas23 for your work and hopefully I get good enough to contribute one day :) .
